File: 0143-reorder-list/0143-reorder-list.py
# Definition for singly-linked list.
# class ListNode:
#     def __init__(self, val=0, next=None):
#         self.val = val
#         self.next = next
class Solution:
    def reorderList(self, head: Optional[ListNode]) -> None:
        """
        Do not return anything, modify head in-place instead.
        """
        slow, fast = head, head.next
        while fast and fast.next:
            slow = slow.next
            fast = fast.next.next
        
        second = slow.next
        slow.next = None
        first = head
        prev = None

        while second:
            tmp = second.next
            second.next = prev
            prev = second
            second = tmp
# prev - reversed second part
# merge prev and first part
        while prev:
            temp1, temp2 = first.next, prev.next
            first.next = prev
            prev.next = temp1
            first , prev = temp1, temp2 
        


        

File: 0143-reorder-list/0143-reorder-list.java
/**
 * Definition for singly-linked list.
 * public class ListNode {
 *     int val;
 *     ListNode next;
 *     ListNode() {}
 *     ListNode(int val) { this.val = val; }
 *     ListNode(int val, ListNode next) { this.val = val; this.next = next; }
 * }
 */
class Solution {
    public void reorderList(ListNode head) {
        ListNode slow = head;
        ListNode fast = head.next;
        while (fast != null && fast.next != null){
            slow = slow.next;
            fast = fast.next.next;
        }
        ListNode second = slow.next;
        slow.next = null;
        ListNode first = head;
        ListNode prev = null; // for reverse second part;

        while (second != null){
            ListNode tmp = second.next;
            second.next = prev;
            prev = second;
            second = tmp;
        }

        // merge reversed second and first

        while (prev != null){
            ListNode tmp1 = first.next;
            ListNode tmp2 = prev.next;

            first.next = prev;
            prev.next = tmp1;

            first = tmp1;
            prev =  tmp2;
        }
     
    }
}